share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -321,16 +342,25 @@ function(ExternalZephyrProject_Cmake)
          "${ZCMAKE_APPLICATION}_BOARD:STRING=${${ZCMAKE_APPLICATION}_BOARD}\n"
     )
   endif()
+  list(APPEND sysbuild_cache_strings
+       "${ZCMAKE_APPLICATION}_KCONFIG_BINARY_DIR:PATH=${${ZCMAKE_APPLICATION}_KCONFIG_BINARY_DIR}\n"
+  )
 
   file(WRITE ${${ZCMAKE_APPLICATION}_CACHE_FILE}.tmp ${sysbuild_cache_strings})
   zephyr_file_copy(${${ZCMAKE_APPLICATION}_CACHE_FILE}.tmp
                    ${${ZCMAKE_APPLICATION}_CACHE_FILE} ONLY_IF_DIFFERENT
   )
 
+  set(dotconfigsysbuild ${${ZCMAKE_APPLICATION}_KCONFIG_BINARY_DIR}/.config.sysbuild)
+  get_target_property(config_content ${ZCMAKE_APPLICATION} CONFIG)
+  string(CONFIGURE "${config_content}" config_content)
+  file(WRITE ${dotconfigsysbuild} ${config_content})
+
   execute_process(
     COMMAND ${CMAKE_COMMAND}
       -G${CMAKE_GENERATOR}
         ${CMAKE_ARGS}
+      -DCONFIG_FORCED:FILEPATH=${dotconfigsysbuild}
```

Comment:
> * Dependency on `${BINARY_DIR}/zephyr/.config`, to re-run sysbuild when image configuration changes.
 
I understand you reasons but extreme care must be taken, cause remember a CMake re-run of sysbuild will trigger a CMake re-run of the image which might result in an updated image `.config`, which then again may trigger a sysbuild CMake re-run if such a dependency is created.
So a risk of each a full CMake re-run before each build.

> * Consistency checks on sysbuild-controlled symbols, performed after `import_kconfig()`.
 
would be nice, but unsure how it will work out in practice, cause a user doing a `<image>_menuconfig` will only trigger a rebuild in that image (and the CMake of that image) and dependents of the image.

However, it might be that the image itself can compare the `.config` and forced `.config` and issue a better warning if there is a mismatch.

> * Only set `CONFIG_FORCED` if sysbuild's own `.config` has changed, so that we can try to report user errors instead of immediately correcting them.

CONFIG_FORCED points to a file, not specific Kconfig setting.
Also it is a cache variable, so wouldn't change anything if it is only provided on first CMake run, or if we provide it on each invocation.

Secondly, we shouldn't report errors and expecting users to manually fix them.
First because they would anyway need to set the value to the CMake enforced value.
Second, because a FATAL_ERROR in CMake prevents users from running menuconfig, hence making it harder to fix such issues, ref: https://github.com/zephyrproject-rtos/zephyr/issues/9573

Reply:
> So a risk of each a full CMake re-run before each build.

That's a good point. So far, I haven't come across this, but I see that it's not preventable, so I retract suggestion (1).

> a user doing a <image>_menuconfig will only trigger a rebuild in that image (and the CMake of that image) and dependents of the image.

That's why I suggested (1) initially, so that sysbuild gets rebuilt as well. Without that, you're right, (2) won't work.

> However, it might be that the image itself can compare the .config and forced .config and issue a better warning if there is a mismatch.

I think that could work too. I assume this would have to be implemented in `kconfig.py`.

> CONFIG_FORCED points to a file, not specific Kconfig setting.

I know, but now that you point it out, it should be renamed to avoid confusing other people.

> Also it is a cache variable, so wouldn't change anything if it is only provided on first CMake run, or if we provide it on each invocation.

Technically, we should be able to pass `-UCONFIG_FORCED` if we want to unset the cache variable, right?

> FATAL_ERROR in CMake prevents users from running menuconfig

True. I've seen this issue and it is an annoyance. We should still print a warning before applying the forced value.

Reply:
> > > CONFIG_FORCED points to a file, not specific Kconfig setting.
> > 
> > 
> > I know, but now that you point it out, it should be renamed to avoid confusing other people.
> 
> that a very good suggestion. Maybe `FORCED_CONF_FILE`, to be aligned with `CONF_FILE` and `EXTRA_CONF_FILE` (see #57994).

Updated to `FORCED_CONF_FILE`.

Reply:
> Guess you missed this comment in top of the PR:

Ah no, that was just poor choice of words on my part. What I had in mind was to print a _better_ warning, as I suggested earlier, since I found the one given by kconfiglib to be a bit small and not self-explanatory in this context.

Reply:
>  What I had in mind was to print a better warning, as I suggested earlier, since I found the one given by kconfiglib to be a bit small and not self-explanatory in this context.

then we are aligned.
Agreed that a better warning would be nice, but we need to be a bit careful cause in principle we want to minimize customization to kconfiglib.py itself. Zephyr specifics should ideally be handled in `kconfig.py`.

Therefore I prefer that we look into improved message as followup to this PR.

Reply:
> Updated to `FORCED_CONF_FILE`.

Nice! Please update the commit message as well.

Reply:
> Nice! Please update the commit message as well.

Done.